core/src/main/scala/kafka/coordinator/transaction/TransactionStateManager.scala

```diff
@@ -108,8 +108,59 @@ class TransactionStateManager(brokerId: Int,
     }
   }
 
-  def enablePidExpiration() {
-    // TODO: add producer id expiration logic
+  def enableTransactionalIdExpiration() {
+    scheduler.schedule("transactionalId-expiration", () => {
+      inLock(stateLock) {
+        val now = time.milliseconds()
+
+        val transactionalIdByPartition: Map[Int, mutable.Iterable[String]] = transactionMetadataCache.flatMap { case (_, entry) =>
+          entry.metadataPerTransactionalId.filter { case (_, txnMetadata) =>
+            txnMetadata.txnLastUpdateTimestamp <= now - config.transactionalIdExpirationMs
+          }.map { case (transactionalId, _) =>
+            transactionalId
+          }
+        }.groupBy { transactionalId =>
+          partitionFor(transactionalId)
+        }
+
+        val recordsPerPartition = transactionalIdByPartition
```

Comment:
For `initPid` if a pending state is set it already will return  `Errors.CONCURRENT_TRANSACTION` which should suffice? I don't see any reason to make it more complicated than that.
I'm not sure about the `addPartition` and `endTxn` cases as how would you get into the state where a Transaction is likely dead and then you try to `addPartition` or `endTxn`? i.e., the transaction will have been aborted long before then if it timed out. So in the `addPartition` case then transaction will have had to have been previously `Empty` in which case it has already been through `initPid` and backed off if there is a pending state defined.